### PR TITLE
Total refactor of the way we handle and draw mutant_bodyparts (+ missed mirrors that motivated me to do this)

### PR DESCRIPTION
**Pull request by GoldenAlpharex**
## About The Pull Request

# After about ~25-30~ 40~ish hours of work, and nearly a year of "I'll do it eventually", here it finally is. The mutant_bodyparts total refactor.

### What Time-Green said in his PR:
Bodypart overlays are now drawn by the new `/datum/bodypart_overlay` datum.

External organs no longer draw anything and instead add a special `/datum/bodypart_overlay/mutant` to the bodypart, which draws everything

Makes it way easier to add custom overlays to limbs, since the whole system is now modularized and external organs are just one implementation of it

I haven't moved anything but external organs to this new system, I'll move eyes, bodymarkings, hair, lipstick etc to this later

New pipeline is as follows:
- External organ added to limb
- External organ adds /datum/bodypart_overlay/mutant to limb to bodypart_overlays
- Limb updates its icon, looks for all /datum/bodypart_overlay in bodypart_overlays
- Very cool new overlay on your limb!

### What I did to follow along:
I had to do something about our mutant bodyparts, because they were just getting far too out of hands. The code for them was genuinely awful. Now *everything* that was in `dna.mutant_bodyparts` was changed to be handled by an external organ, so that it wouldn't just be sprites floating out of nowhere and nigh-impossible to keep track of, or modify. They're now using the upstream rendering pipeline that can be seen above, with some modifications for our sake to make multi-colored bodyparts be rendered properly, as well as emissives and MODsuit overlays (which I hope we can make a lot less baked in, in the future).

Now it's far more manageable, and I finally broke away from `handle_mutant_bodyparts()`, entirely. It doesn't do anything anymore. Thank me later.

All of the core of the changes can be found in `modular_skyrat/master_files/code/datums/bodypart_overlays/mutant_bodypart_overlay.dm`. If I can be assed, in the future, I'll try to regroup a lot of the rendering stuff together so it's easier to find, but that's a job for another time.

I've got to take the changes to ears (making them an external organ instead) upstream, because it's not the best thing in the world at the moment, having two organs for it, but it was just so we would have something functioning in the meantime.

There's still some dead code lying around, and a lot of places where things still aren't done in the most efficient ways. A lot of useless redraws are still around, and we'll have to get together to try and chase those down.

There's also some optimization left to do, namely on the way we choose the layers to then try to add to the mobs. I think we'll probably be able to cut down about 25% of the overlays that are there on average, to account for all the places where the `relevant_layers` aren't set properly. Which I'll need help to track, because I don't use most of these mutant bodyparts myself. I also need to actually properly use `relevant_layers` in the mutant bodypart overlays themselves, to avoid drawing empty overlays that are completely pointless and useless. In any case, I'll need some help if we want to see this through, but it'll be worth it, trust me.

There's one instance where I made a non-modular edit without mentioning it, but it's broken code upstream that I'm about to PR there anyway, so I'll deal with that conflict when it trickles down here.

I'll be happy to go into more details of how this all works if the desire is there, I've tried to document my code as much as possible as I went along, so it made sense, let me know if anything needs to be clearer though.

## How This Contributes To The Skyrat Roleplay Experience
Performances of organ-rendering-related code should be better. This isn't perfect, there's no caching of the overlays themselves yet because it hasn't been implemented upstream, I'll probably go and implement it upstream later, if someone else doesn't beat me to it.

Things should also hopefully start working natively a lot better, like changelings or even possibly Genetics, where we'll be able to actually obtain those bodyparts in-round.

Oh, yeah, and many of the organs use placeholder icons, because I didn't have any icons on hand. If anyone's feeling in a spriting mood, feel free to hit me up about it!

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>
  
It works.
image

MODsuits also work. :)
image

</details>

## Changelog

:cl: GoldenAlpharex, Time-Green and lizardqueenlexi
fix: The following are the /tg/ changelogs:
refactor: External organs have been near-completely refactored.
admin: Admin-spawned external organs will load with a random icon and color
fix: fixes angel wings not working for non-humans (it was so fucking broken)
fix: fixes external organs being invisible if they werent initialized with a human
refactor: Removed mutant_bodyparts.dmi and distributed its sprites in a more sensible and granular manner.
code: Now this is the part Golden worked on:
refactor: Completely refactored the way mutant_bodyparts were handled. To make a long story short, they're all external organs now, some of them can be removed when it makes sense, and they should render a lot better from now on.
fix: There should already be a lot less overlays on your character at once, with some more optimizations regarding that coming soon.
fix: Functional wings obtained by the flight potion should now display properly when opened and closed. Synths should also have access to their robotic wings once again.
qol: All of the contents of the "Body Markings" option in the Appearance tab were moved to the Augments+ tab, in the Markings section, for more consistency.
fix: Synth eyes no longer show on their sprite when they have the IPC screen on.
fix: At long last, podpeople hair works now, woo!
fix: Fixed an issue where the shapeshifting action was accidentally messing up the global sprite accessory list, making some options unavailable on accident.
fix: If you notice any bugs related to the rendering of your characters, please report them using the "Report Issue" button at the top-right corner of the screen, mentioning #19635 in it for ease of reference.
/:cl:

### CODE REVIEW

File: modular_skyrat/modules/customization/modules/mob/dead/new_player/sprite_accessories/moth_antennae.dm [hunk 1]
```diff
@@ -3,34 +3,26 @@
...
-		return TRUE
-//	Hide accessory if flagged to do so
+	// Hide accessory if flagged to do so
 	if((wearer.head.flags_inv & HIDEHAIR || (wearer.wear_mask && (wearer.wear_mask.flags_inv & HIDEHAIR))) \
 		&& !(wearer.head.flags_inv & SHOWSPRITEEARS || wearer.wear_mask.flags_inv & SHOWSPRITEEARS))
-		return FALSE
+		return TRUE
 
+/*
+/datum/bodypart_overlay/mutant/antennae/can_draw_on_bodypart(mob/living/carbon/human/wearer)
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/customization/modules/mob/dead/new_player/sprite_accessories/moth_antennae.dm]: what here

File: modular_skyrat/modules/customization/modules/surgery/organs/moth_antennae.dm [hunk 2]
```diff
@@ -0,0 +1,3 @@
+/obj/item/organ/external/antennae
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/surgery/organs/moth_antennae.dm:1]: Seems like I forgot to tick this one, but it was still working without it. Somehow.

**Comment by Gandalf2k15** [comment]
https://github.com/Skyrat-SS13/Skyrat-tg/pull/19660 - LINE 16 UNCOMMENT ONCE MERGED.

File: code/datums/bodypart_overlays/bodypart_overlay.dm [hunk 3]
```diff
@@ -0,0 +1,82 @@
+///Bodypart ovarlay datum. These can be added to any limb to give them a proper overlay, that'll even stay if the limb gets removed
+///This is the abstract parent, don't use it!!
+/datum/bodypart_overlay
+	///Sometimes we need multiple layers, for like the back, middle and front of the person (EXTERNAL_FRONT, EXTERNAL_ADJACENT, EXTERNAL_BEHIND)
+	var/layers
+	///List of all possible layers. Used for looping through in drawing
+	var/static/list/all_layers = list(EXTERNAL_FRONT, EXTERNAL_ADJACENT, EXTERNAL_BEHIND, EXTERNAL_FRONT_UNDER_CLOTHES, EXTERNAL_FRONT_OVER) // SKYRAT EDIT - Customization - ORIGINAL: var/static/list/all_layers = list(EXTERNAL_FRONT, EXTERNAL_ADJACENT, EXTERNAL_BEHIND)
```
Comment (Gandalf2k15) [inline code/datums/bodypart_overlays/bodypart_overlay.dm:7]: Skyrat comment? What happened to this file?

File: code/datums/bodypart_overlays/bodypart_overlay.dm [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+///Converts a bitflag to the right layer. I'd love to make this a static index list, but byond made an attempt on my life when i did
+/datum/bodypart_overlay/proc/bitflag_to_layer(layer)
+	switch(layer)
+		if(EXTERNAL_BEHIND)
+			return BODY_BEHIND_LAYER
+		if(EXTERNAL_ADJACENT)
+			return BODY_ADJ_LAYER
+		if(EXTERNAL_FRONT)
+			return BODY_FRONT_LAYER
+		//SKYRAT EDIT ADDITION - Customization
```
Comment (Gandalf2k15) [inline code/datums/bodypart_overlays/bodypart_overlay.dm:65]: ditto...?

File: code/datums/bodypart_overlays/mutant_bodypart_overlay.dm [hunk 5]
```diff
@@ -0,0 +1,116 @@
+///Variant of bodypart_overlay meant to work synchronously with external organs. Gets imprinted upon Insert in on_species_gain
```
Comment (Gandalf2k15) [inline code/datums/bodypart_overlays/mutant_bodypart_overlay.dm:1]: Were these files straight up missing?

File: code/datums/dna.dm [hunk 6]
```diff
@@ -675,7 +675,7 @@ GLOBAL_LIST_INIT(total_uf_len_by_block, populate_total_uf_len_by_block())
 	if(dna.features["pod_hair"])
 		dna.features["pod_hair"] = GLOB.pod_hair_list[deconstruct_block(get_uni_feature_block(features, DNA_POD_HAIR_BLOCK), GLOB.pod_hair_list.len)]
 
-	for(var/obj/item/organ/external/external_organ as anything in external_organs)
```
Comment (Gandalf2k15) [inline code/datums/dna.dm:678]: changes to commented out code

File: code/modules/client/preferences/species_features/lizard.dm [hunk 7]
```diff
@@ -8,7 +8,7 @@
 	lizard.Blend(eyes, ICON_OVERLAY)
 
 	if (include_snout)
-		lizard.Blend(icon('icons/mob/species/mutant_bodyparts.dmi', "m_snout_round_ADJ", EAST), ICON_OVERLAY)
```
Comment (Gandalf2k15) [inline code/modules/client/preferences/species_features/lizard.dm:11]: commented out changes

File: code/modules/mob/living/carbon/human/dummy.dm [hunk 8]
```diff
@@ -95,10 +96,13 @@ INITIALIZE_IMMEDIATE(/mob/living/carbon/human/dummy)
 	target.dna.features["wings"] = "None" // SKYRAT EDIT CHANGE - Customization
```
Comment (Gandalf2k15) [inline code/modules/mob/living/carbon/human/dummy.dm]: remove skyrat edit comment

File: modular_skyrat/master_files/code/modules/client/preferences/middleware/limbs_and_markings.dm [hunk 9]
```diff
@@ -2,7 +2,7 @@
 #define LIMBS_DEFAULT_NAME "None"
 
 /datum/preference_middleware/limbs_and_markings/post_set_preference(mob/user, preference, value)
-	preferences.character_preview_view.update_body()
+	// preferences.character_preview_view.update_body()
```
Comment (Gandalf2k15) [inline modular_skyrat/master_files/code/modules/client/preferences/middleware/limbs_and_markings.dm]: ? explain why removed or just remove fully

File: modular_skyrat/master_files/code/modules/client/preferences/mutant_parts.dm [hunk 10]
```diff
@@ -68,18 +68,34 @@
 /datum/preference/toggle/eye_emissives/proc/is_allowed(datum/preferences/preferences)
 	return preferences.read_preference(/datum/preference/toggle/allow_emissives)
 
-// Body Markings
+// Body Markings - This isn't used anymore and thus I'm making it not do anything.
```
Comment (Gandalf2k15) [inline modular_skyrat/master_files/code/modules/client/preferences/mutant_parts.dm:71]: Then remove it fully from the code?

**Review by Gandalf2k15 — CHANGES_REQUESTED** [review]
looks good to me, but ideally we want to also atomise the heck out of the customisation module.

File: code/datums/bodypart_overlays/bodypart_overlay.dm (hunk 3 above)
Comment (GoldenAlpharex) [reply code/datums/bodypart_overlays/bodypart_overlay.dm:7]: This was in the missed mirror, and I then added some changes myself.

File: code/datums/bodypart_overlays/bodypart_overlay.dm (hunk 4 above)
Comment (GoldenAlpharex) [reply code/datums/bodypart_overlays/bodypart_overlay.dm:65]: This was in the missed mirror, and I then added some changes myself.

File: code/datums/bodypart_overlays/mutant_bodypart_overlay.dm (hunk 5 above)
Comment (GoldenAlpharex) [reply code/datums/bodypart_overlays/mutant_bodypart_overlay.dm:1]: Yes, they were in the missed mirror about the bodypart overlays system, created by Time-Green.

File: code/datums/dna.dm (hunk 6 above)
Comment (GoldenAlpharex) [reply code/datums/dna.dm:678]: From the mirror.

File: code/modules/client/preferences/species_features/lizard.dm (hunk 7 above)
Comment (GoldenAlpharex) [reply code/modules/client/preferences/species_features/lizard.dm:11]: From the mirror.

File: modular_skyrat/master_files/code/modules/client/preferences/mutant_parts.dm (hunk 10 above)
Comment (GoldenAlpharex) [reply modular_skyrat/master_files/code/modules/client/preferences/mutant_parts.dm:71]: I didn't want to remove it yet, just in the off-chance that anything came up during the test-merges. I'll be removing it later on.

Also, having it removed in a subsequent PR will make it easier to reference in the future.

File: modular_skyrat/master_files/code/modules/client/preferences/middleware/limbs_and_markings.dm (hunk 9 above)
Comment (GoldenAlpharex) [reply modular_skyrat/master_files/code/modules/client/preferences/middleware/limbs_and_markings.dm]: I forgot to remove it fully, my bad, good catch.

**Comment by avimour** [comment]
Tail wags seem to be gone, or at least for the xenomorph tail!

**Comment by Triple-Skrek** [comment]
The eyes on the bug snout now take their colour from mutant colour 1 instead of 3. Would be very nice to separate those two again.

**Comment by nikothedude** [comment]
Synthetic lizard - snout tertiary now uses secondary colors for the visor instead of tertiary colors. Not a bad change just, a change that happened that made my lizard look weird.

**Comment by GoldenAlpharex** [comment]
> Tail wags seem to be gone, or at least for the xenomorph tail!

Going to be fixed when I push the change, thanks for the report!

**Comment by BriareosBlue** [comment]
I suspect something about this might have brought back a return of helmets and masks no longer using their snouted versions on characters with snouts.

**Comment by GoldenAlpharex** [comment]
> I suspect something about this might have brought back a return of helmets and masks no longer using their snouted versions on characters with snouts.

Is that something that started happening yesterday, or was it an issue before that?
Do you have any helmet/mask that you could give me the name of so I can reproduce that issue?

**Comment by GoldenAlpharex** [comment]
> > I suspect something about this might have brought back a return of helmets and masks no longer using their snouted versions on characters with snouts.
> 
> Is that something that started happening yesterday, or was it an issue before that? Do you have any helmet/mask that you could give me the name of so I can reproduce that issue?

That issue has been fixed.

**Comment by Nerev4r** [comment]
image
Plasmaman limbs now give you an evil invisible arm instead of an evil skeleton arm.

**Comment by GoldenAlpharex** [comment]
> image Plasmaman limbs now give you an evil invisible arm instead of an evil skeleton arm.

Also fixed

**Review by Gandalf2k15 — APPROVED** [review]

**Pull request merged**